### Sentiment Analysis through Integer Encoding Method

<h4>How we do integer encoding</h4>

In [46]:
import numpy as np

#input data contains 10 sentences
docs = ['go india',
		'india india',
		'hip hip hurray',
		'jeetega bhai jeetega india jeetega',
		'bharat mata ki jai',
		'kohli kohli',
		'sachin sachin',
		'dhoni dhoni',
		'modi ji ki jai',
		'inquilab zindabad']

In [47]:
from tensorflow.keras.preprocessing.text import Tokenizer

tokenizer = Tokenizer(oov_token='<nothing>') 
#oov_token stands for Out-Of-Vocabulary token.
# It is a special token used by the tokenizer to represent words that were not seen
#during training (i.e., words that are not in the tokenizer's vocabulary).

In [48]:
tokenizer.fit_on_texts(docs)

In [49]:
tokenizer.word_index

{'<nothing>': 1,
 'india': 2,
 'jeetega': 3,
 'hip': 4,
 'ki': 5,
 'jai': 6,
 'kohli': 7,
 'sachin': 8,
 'dhoni': 9,
 'go': 10,
 'hurray': 11,
 'bhai': 12,
 'bharat': 13,
 'mata': 14,
 'modi': 15,
 'ji': 16,
 'inquilab': 17,
 'zindabad': 18}

In [50]:
tokenizer.word_counts #kon sa word kitni baar aya

OrderedDict([('go', 1),
             ('india', 4),
             ('hip', 2),
             ('hurray', 1),
             ('jeetega', 3),
             ('bhai', 1),
             ('bharat', 1),
             ('mata', 1),
             ('ki', 2),
             ('jai', 2),
             ('kohli', 2),
             ('sachin', 2),
             ('dhoni', 2),
             ('modi', 1),
             ('ji', 1),
             ('inquilab', 1),
             ('zindabad', 1)])

In [51]:
tokenizer.document_count #how many sentences

10

In [52]:
sequences = tokenizer.texts_to_sequences(docs)
sequences

[[10, 2],
 [2, 2],
 [4, 4, 11],
 [3, 12, 3, 2, 3],
 [13, 14, 5, 6],
 [7, 7],
 [8, 8],
 [9, 9],
 [15, 16, 5, 6],
 [17, 18]]

In [53]:
from keras.utils import pad_sequences #applying padding

In [54]:
sequences = pad_sequences(sequences,padding='post')

In [55]:
sequences

array([[10,  2,  0,  0,  0],
       [ 2,  2,  0,  0,  0],
       [ 4,  4, 11,  0,  0],
       [ 3, 12,  3,  2,  3],
       [13, 14,  5,  6,  0],
       [ 7,  7,  0,  0,  0],
       [ 8,  8,  0,  0,  0],
       [ 9,  9,  0,  0,  0],
       [15, 16,  5,  6,  0],
       [17, 18,  0,  0,  0]], dtype=int32)

<h4>Applying on Dataset imbd</h4>

In [56]:
from keras.datasets import imdb
from keras import Sequential
from keras.layers import Dense,SimpleRNN,Embedding,Flatten

In [57]:
(X_train,y_train),(X_test,y_test) = imdb.load_data()

In [58]:
X_train.shape

(25000,)

In [59]:
X_test.shape

(25000,)

In [60]:
X_train #in vocabulary we have assigned number to each words thats why we are seeing numbers

array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 22665, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 21631, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 19193, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 10311, 8, 4, 107, 117, 5952, 15, 256, 4, 31050, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 12118, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
       list([1, 194, 1

In [61]:
X_train[0] #this data is already integer encoded and already preprocessed

[1,
 14,
 22,
 16,
 43,
 530,
 973,
 1622,
 1385,
 65,
 458,
 4468,
 66,
 3941,
 4,
 173,
 36,
 256,
 5,
 25,
 100,
 43,
 838,
 112,
 50,
 670,
 22665,
 9,
 35,
 480,
 284,
 5,
 150,
 4,
 172,
 112,
 167,
 21631,
 336,
 385,
 39,
 4,
 172,
 4536,
 1111,
 17,
 546,
 38,
 13,
 447,
 4,
 192,
 50,
 16,
 6,
 147,
 2025,
 19,
 14,
 22,
 4,
 1920,
 4613,
 469,
 4,
 22,
 71,
 87,
 12,
 16,
 43,
 530,
 38,
 76,
 15,
 13,
 1247,
 4,
 22,
 17,
 515,
 17,
 12,
 16,
 626,
 18,
 19193,
 5,
 62,
 386,
 12,
 8,
 316,
 8,
 106,
 5,
 4,
 2223,
 5244,
 16,
 480,
 66,
 3785,
 33,
 4,
 130,
 12,
 16,
 38,
 619,
 5,
 25,
 124,
 51,
 36,
 135,
 48,
 25,
 1415,
 33,
 6,
 22,
 12,
 215,
 28,
 77,
 52,
 5,
 14,
 407,
 16,
 82,
 10311,
 8,
 4,
 107,
 117,
 5952,
 15,
 256,
 4,
 31050,
 7,
 3766,
 5,
 723,
 36,
 71,
 43,
 530,
 476,
 26,
 400,
 317,
 46,
 7,
 4,
 12118,
 1029,
 13,
 104,
 88,
 4,
 381,
 15,
 297,
 98,
 32,
 2071,
 56,
 26,
 141,
 6,
 194,
 7486,
 18,
 4,
 226,
 22,
 21,
 134,
 476,
 26,
 480,
 5

In [62]:
len(X_train[2])

141

In [71]:
from tensorflow.keras.preprocessing.sequence import pad_sequences

maxlen = 200

X_train = pad_sequences(X_train, maxlen=maxlen)
X_test = pad_sequences(X_test, maxlen=maxlen)

print(X_train.dtype)
print(X_train.shape)

int32
(25000, 200)


In [72]:
X_train[0]

array([    5,    25,   100,    43,   838,   112,    50,   670, 22665,
           9,    35,   480,   284,     5,   150,     4,   172,   112,
         167, 21631,   336,   385,    39,     4,   172,  4536,  1111,
          17,   546,    38,    13,   447,     4,   192,    50,    16,
           6,   147,  2025,    19,    14,    22,     4,  1920,  4613,
         469,     4,    22,    71,    87,    12,    16,    43,   530,
          38,    76,    15,    13,  1247,     4,    22,    17,   515,
          17,    12,    16,   626,    18, 19193,     5,    62,   386,
          12,     8,   316,     8,   106,     5,     4,  2223,  5244,
          16,   480,    66,  3785,    33,     4,   130,    12,    16,
          38,   619,     5,    25,   124,    51,    36,   135,    48,
          25,  1415,    33,     6,    22,    12,   215,    28,    77,
          52,     5,    14,   407,    16,    82, 10311,     8,     4,
         107,   117,  5952,    15,   256,     4, 31050,     7,  3766,
           5,   723,

In [75]:
model = Sequential()

model.add(SimpleRNN(32,input_shape=(200,1),return_sequences=False)) #(200->timestep,1 feature at a time)
model.add(Dense(1,activation='sigmoid'))

model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ simple_rnn_2 (SimpleRNN)             │ (None, 32)                  │           1,088 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 1)                   │              33 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 1,121 (4.38 KB)

 Trainable params: 1,121 (4.38 KB)

 Non-trainable params: 0 (0.00 B)

In [76]:
model.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])

model.fit(X_train,y_train,epochs=5,validation_data=(X_test,y_test))

Epoch 1/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 43s 52ms/step - accuracy: 0.5050 - loss: 0.6959 - val_accuracy: 0.5001 - val_loss: 0.6935
Epoch 2/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 39s 50ms/step - accuracy: 0.5062 - loss: 0.6934 - val_accuracy: 0.5074 - val_loss: 0.6936
Epoch 3/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 40s 51ms/step - accuracy: 0.5082 - loss: 0.6933 - val_accuracy: 0.5010 - val_loss: 0.6943
Epoch 4/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 41s 52ms/step - accuracy: 0.5052 - loss: 0.6932 - val_accuracy: 0.5000 - val_loss: 0.6932
Epoch 5/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 43s 54ms/step - accuracy: 0.5084 - loss: 0.6929 - val_accuracy: 0.5016 - val_loss: 0.6941


<h4>Testing the performance of model by predicting something with it</h4>

In [81]:

word_index = imdb.get_word_index()

# Create a reverse word index to map words back to integers
# Remember that the IMDB dataset reserves indices 0, 1, and 2 for padding, start of sentence, and unknown words.
# So we need to add 3 to each word index.
word_to_id = {k: v + 3 for k, v in word_index.items()}
id_to_word = {value: key for key, value in word_to_id.items()}

# Define a function to preprocess a new review
def preprocess_review(review_text, word_to_id, maxlen=50):
    # Convert review to lowercase and split into words
    words = review_text.lower().split()

    # Convert words to their corresponding IMDB indices
    # Handle words not in the vocabulary by mapping them to index 2 (unknown)
    indexed_review = [word_to_id.get(word, 2) for word in words]

    # Pad the sequence to the fixed length (maxlen=50)
    padded_review = pad_sequences([indexed_review], padding='post', maxlen=maxlen)

    # Reshape for the SimpleRNN input layer (batch_size, timesteps, features)
    # The model was built with input_shape=(50,1), so we add a feature dimension
    reshaped_review = np.expand_dims(padded_review, axis=-1)

    return reshaped_review


# Example new negative review
negative_review = "This film was terrible, a complete waste of time. I would not recommend it at all."


processed_negative_review = preprocess_review(negative_review, word_to_id, maxlen=50)

# Predict the sentiment for the negative review
negative_prediction = model.predict(processed_negative_review)

# Interpret the prediction
if negative_prediction[0][0] > 0.5:
    negative_sentiment = "Positive"
else:
    negative_sentiment = "Negative"

print(f"Review: '{negative_review}'")
print(f"Predicted sentiment score: {negative_prediction[0][0]:.4f}")
print(f"Predicted sentiment: {negative_sentiment}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 75ms/step
Review: 'This film was terrible, a complete waste of time. I would not recommend it at all.'
Predicted sentiment score: 0.4145
Predicted sentiment: Negative


In [ ]:
#the goal is not to achieve best model the goal is to learn how to implement integer encodingvectorisation and also how to use it for sentimment analysis

# Embedding


In [82]:
docs = ['go india',
		'india india',
		'hip hip hurray',
		'jeetega bhai jeetega india jeetega',
		'bharat mata ki jai',
		'kohli kohli',
		'sachin sachin',
		'dhoni dhoni',
		'modi ji ki jai',
		'inquilab zindabad']

In [84]:
from tensorflow.keras.preprocessing.text import Tokenizer

tokenizer = Tokenizer() 

In [85]:
tokenizer.fit_on_texts(docs)

In [86]:
len(tokenizer.word_index)

17

In [87]:
sequences = tokenizer.texts_to_sequences(docs)
sequences

[[9, 1],
 [1, 1],
 [3, 3, 10],
 [2, 11, 2, 1, 2],
 [12, 13, 4, 5],
 [6, 6],
 [7, 7],
 [8, 8],
 [14, 15, 4, 5],
 [16, 17]]

In [88]:
from keras.utils import pad_sequences
sequences = pad_sequences(sequences,padding='post')
sequences

array([[ 9,  1,  0,  0,  0],
       [ 1,  1,  0,  0,  0],
       [ 3,  3, 10,  0,  0],
       [ 2, 11,  2,  1,  2],
       [12, 13,  4,  5,  0],
       [ 6,  6,  0,  0,  0],
       [ 7,  7,  0,  0,  0],
       [ 8,  8,  0,  0,  0],
       [14, 15,  4,  5,  0],
       [16, 17,  0,  0,  0]], dtype=int32)

In [92]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding

model = Sequential()
model.add(Embedding(input_dim=17, output_dim=2))

model.summary()

Model: "sequential_5"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_2 (Embedding)              │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

In [96]:
model = Sequential()
model.add(Embedding(18,output_dim=2,input_length=5)) #in embedding we first tell vocab_words+1,output_dim=2 ka matlab hai that we want to represent each wordwith two numbers, input length_ you have to tell how many input words you have 

model.summary()

Model: "sequential_7"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_4 (Embedding)              │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

In [97]:
pred = model.predict(sequences)
print(pred)

#here we are getting two numbers for every word as we have classified in output_dim

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 238ms/step
[[[ 0.02322591 -0.03819131]
  [ 0.0202809   0.04499117]
  [ 0.02513209  0.00738934]
  [ 0.02513209  0.00738934]
  [ 0.02513209  0.00738934]]

 [[ 0.0202809   0.04499117]
  [ 0.0202809   0.04499117]
  [ 0.02513209  0.00738934]
  [ 0.02513209  0.00738934]
  [ 0.02513209  0.00738934]]

 [[ 0.01278981  0.03242991]
  [ 0.01278981  0.03242991]
  [-0.00116154  0.00308703]
  [ 0.02513209  0.00738934]
  [ 0.02513209  0.00738934]]

 [[-0.01623375  0.03161405]
  [ 0.0126015  -0.03742069]
  [-0.01623375  0.03161405]
  [ 0.0202809   0.04499117]
  [-0.01623375  0.03161405]]

 [[-0.04879481 -0.0452772 ]
  [ 0.01647389 -0.00955172]
  [-0.01088238 -0.03210341]
  [ 0.01444763  0.0050164 ]
  [ 0.02513209  0.00738934]]

 [[ 0.02806417  0.01655019]
  [ 0.02806417  0.01655019]
  [ 0.02513209  0.00738934]
  [ 0.02513209  0.00738934]
  [ 0.02513209  0.00738934]]

 [[ 0.04398782  0.03465278]
  [ 0.04398782  0.03465278]
  [ 0.02513209  0.00738934]
  [ 0.02513209  0.0073893

## now sentiment analysis with embedding

In [105]:
from tensorflow.keras.datasets import imdb

(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=10000)

In [106]:
X_train = pad_sequences(X_train,padding='post',maxlen=50)
X_test = pad_sequences(X_test,padding='post',maxlen=50)

In [107]:
X_train.shape

(25000, 50)

In [108]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, Dense

model = Sequential()

model.add(Embedding(
    input_dim=10000,
    output_dim=2,
    input_length=50   # or just omit it
))

model.add(SimpleRNN(32, return_sequences=False))
model.add(Dense(1, activation='sigmoid'))

model.summary()

Model: "sequential_10"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_7 (Embedding)              │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ simple_rnn_4 (SimpleRNN)             │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_4 (Dense)                      │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

In [110]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['acc'])
history = model.fit(X_train, y_train,epochs=5,validation_data=(X_test,y_test))

Epoch 1/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 19s 20ms/step - acc: 0.9087 - loss: 0.2389 - val_acc: 0.7977 - val_loss: 0.4956
Epoch 2/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 15s 19ms/step - acc: 0.9218 - loss: 0.2138 - val_acc: 0.7897 - val_loss: 0.5434
Epoch 3/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 15s 20ms/step - acc: 0.9324 - loss: 0.1903 - val_acc: 0.7819 - val_loss: 0.6106
Epoch 4/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 14s 18ms/step - acc: 0.9401 - loss: 0.1705 - val_acc: 0.7788 - val_loss: 0.6220
Epoch 5/5
782/782 ━━━━━━━━━━━━━━━━━━━━ 15s 19ms/step - acc: 0.9432 - loss: 0.1609 - val_acc: 0.7736 - val_loss: 0.6349


### Prediction

In [111]:

word_index = imdb.get_word_index()


word_to_id = {k: v + 3 for k, v in word_index.items()}
id_to_word = {value: key for key, value in word_to_id.items()}

# Define a function to preprocess a new review
def preprocess_review(review_text, word_to_id, maxlen=50):
    # Convert review to lowercase and split into words
    words = review_text.lower().split()

    # Convert words to their corresponding IMDB indices
    # Handle words not in the vocabulary by mapping them to index 2 (unknown)
    indexed_review = [word_to_id.get(word, 2) for word in words]

    # Pad the sequence to the fixed length (maxlen=50)
    padded_review = pad_sequences([indexed_review], padding='post', maxlen=maxlen)

    # Reshape for the SimpleRNN input layer (batch_size, timesteps, features)
    # The model was built with input_shape=(50,1), so we add a feature dimension
    reshaped_review = np.expand_dims(padded_review, axis=-1)

    return reshaped_review

# Example new positive review
positive_review = "This movie was an absolute masterpiece! I loved every single moment, truly captivating."

# Example new negative review
negative_review = "This film was terrible, a complete waste of time. I would not recommend it at all."

# Preprocess the positive review
processed_positive_review = preprocess_review(positive_review, word_to_id, maxlen=50)

# Predict the sentiment for the positive review
positive_prediction = model.predict(processed_positive_review)

# Interpret the prediction (sigmoid output: > 0.5 for positive)
if positive_prediction[0][0] > 0.5:
    positive_sentiment = "Positive"
else:
    positive_sentiment = "Negative"

print(f"Review: '{positive_review}'")
print(f"Predicted sentiment score: {positive_prediction[0][0]:.4f}")
print(f"Predicted sentiment: {positive_sentiment}\n")

# Preprocess the negative review
processed_negative_review = preprocess_review(negative_review, word_to_id, maxlen=50)

# Predict the sentiment for the negative review
negative_prediction = model.predict(processed_negative_review)

# Interpret the prediction
if negative_prediction[0][0] > 0.5:
    negative_sentiment = "Positive"
else:
    negative_sentiment = "Negative"

print(f"Review: '{negative_review}'")
print(f"Predicted sentiment score: {negative_prediction[0][0]:.4f}")
print(f"Predicted sentiment: {negative_sentiment}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 278ms/step
Review: 'This movie was an absolute masterpiece! I loved every single moment, truly captivating.'
Predicted sentiment score: 0.9009
Predicted sentiment: Positive

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 66ms/step
Review: 'This film was terrible, a complete waste of time. I would not recommend it at all.'
Predicted sentiment score: 0.0538
Predicted sentiment: Negative


**we can clearly see embedding workd better because it can also capture semantic meaning of words thats why accuracy has increased from approx 50 % to 94 %**